src/Compilers/CSharp/Portable/Declarations/MergedNamespaceDeclaration.cs

```diff
@@ -139,43 +140,100 @@ private ImmutableArray<MergedNamespaceOrTypeDeclaration> MakeChildren()
 
             var children = ArrayBuilder<MergedNamespaceOrTypeDeclaration>.GetInstance();
 
-            if (namespaces != null)
+            addNamespacesToChildren(namespaces, allNamespacesHaveSameName, children);
+            addTypesToChildren(types, allTypesHaveSameIdentity, children);
+
+            return children.ToImmutableAndFree();
+
+            static void addNamespacesToChildren(ArrayBuilder<SingleNamespaceDeclaration> namespaces, bool allNamespacesHaveSameName, ArrayBuilder<MergedNamespaceOrTypeDeclaration> children)
             {
-                if (allNamespacesHaveSameName)
-                {
-                    children.Add(MergedNamespaceDeclaration.Create(namespaces.ToImmutableAndFree()));
-                }
-                else
+                if (namespaces != null)
                 {
-                    var namespaceGroups = namespaces.ToDictionary(n => n.Name, StringOrdinalComparer.Instance);
-                    namespaces.Free();
-
-                    foreach (var namespaceGroup in namespaceGroups.Values)
+                    if (allNamespacesHaveSameName)
                     {
-                        children.Add(MergedNamespaceDeclaration.Create(namespaceGroup));
+                        children.Add(MergedNamespaceDeclaration.Create(namespaces.ToImmutableAndFree()));
+                    }
+                    else
+                    {
+                        // PERF: Don't use ArrayBuilder.ToDictionary directly as it requires an extra dictionary allocation. Other options such
+                        // as MultiDictionary<string, SingleNamespaceDeclaration> and Dictionary<string, OneOrMany<SingleNamespaceDeclaration>>
+                        // are even less appealing as they don't perform well when their value sets grow to contain a large number of items,
+                        // as typically happens when processing the namespaces.
+                        var namespaceGroups = new Dictionary<string, ArrayBuilder<SingleNamespaceDeclaration>>(StringOrdinalComparer.Instance);
```

Comment:
we don't want a pooled dictionary here?  like, have your own pool and get/return from that?

Reply:
I'd like to wait on this, as the only dictionaries that are showing in the profile are the ones for types. Once the test insertion speedometer run is completed, if the dictionary allocations show up, I'll go ahead and make a pool.

Reply:
Not going to add a pool as the speedometer run with these changes didn't indicate the dictionary allocations for the namespaces is a problem.